ska_dbi/sqsh.py

```diff
@@ -0,0 +1,168 @@
+
+        if not Path(SYBASE).exists():
+            raise RuntimeError(f"SYBASE does not exist: {SYBASE}")
+        if not Path(SQSH_BIN).exists():
+            raise RuntimeError(f"sqsh.bin does not exist: {SQSH_BIN}")
+
+        if self.passwd is None and self.sqshrc is None:
+            sqshrc_file = (
+                Path(authdir) / f"sqsh-{self.server}-{self.database}-{self.user}"
+            )
+            if sqshrc_file.exists():
+                self.sqshrc = sqshrc_file
+            else:
+                raise NoPasswordError(
+                    f"None supplied and unable to read inferred sqshrc file {sqshrc_file}"
+                )
+
+    def __enter__(self):
+        """Context manager enter runtime context.  No action required, just return self."""
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """
+        Context manager exit run time context.
+        """
+        pass
+
+
+    def fetch(self, query):
+        """
+        Execute a query and return all returned sql rows a list of lines.
+
+        Note that this doesn't do anything fancy with the types of the return columns.
+        Python Sybase, for example, handled Sybase datetime columns as Python datetimes types
+        and here they are just strings.
+
+        Parameters
+        ----------
+        query : str
+            SQL query to execute (select statements are expected for our use cases)
+
+        Returns
+        -------
+        list of str
+        """
+        envs = ENV_VARS
+        if self.sqshrc is not None:
+            envs = envs + [f"SQSHRC={self.sqshrc}"]
+        cmd = [
+            SQSH_BIN,
+            "-X",
+            "-S",
+            self.server,
+            "-U",
+            self.user,
+            "-D",
+            self.database,
+            "-m",
+            "csv",
+            "-C",
+            query,
+        ]
+        if self.passwd is not None:
+            cmd += ["-P", self.passwd]
+        cmd = ["env"] + envs + cmd
+
+        proc = subprocess.Popen(
+            cmd, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE
+        )
+        stdout, stderr = proc.communicate()
+        if proc.returncode != 0:
+            raise RuntimeError(
+                "sqsh failed with return code %d: %s" % (proc.returncode, stderr)
+            )
+
+        outlines = stdout.decode().splitlines()
+        return outlines
+
+
+    def fetchall(self, query):
+        """
+        Fetches all the rows returned by the query.
+
+        Parameters
+        ----------
+        query : str
+            The SQL query to execute.
+
+        Returns
+        -------
+        astropy.table.Table or None
+            The table containing the fetched rows, or None if no rows are returned.
+        """
+        outlines = self.fetch(query)
+        # Sqsh should always be returning a header line -- return None if that's it.
+        if len(outlines) <= 1:
```

Comment:
Why can't you return a zero-length table here? The types will be wrong (all int64), but otherwise the downstream code should generally work.

This highlights one issue that we are relying on astropy to guess the types instead of asking sqsh for information about the table. For a general purpose interface doing this might be a requirement, but maybe if it just used in a few places with specific tables it is OK to guess.

Reply:
Why do I want to return a zero-length table there?

Reply:
For API consistency and easy of downstream code (so you don't special-case `None` return). Normally any table filter (e.g. database query) with no matches returns a zero-length table. Isn't that what it `fetch` currently does?

Reply:
OK, if the intent is API consistency with DBI.fetchall that's fine.  I didn't care and found it just as easy to handle Nones in mica which was the use case. So keep going back and forth on if there are general use cases and if we have any requirements in API consistency.